### ISSUE TITLE
Do not show error CS0075 on non-ambiguous declarations

### ISSUE THREAD

**Issue by ChrisBertrandIndus** [dotnet/roslyn#56604]
Can we get rid of this error on non-ambiguous declarations ?
So to prevent developers to add parenthesis that make code less readable.

Example 1 :

	int i = (System.Int32) -4;

Example 2 :

	Distance d = (Distance) -25.0;	

	public struct Distance
	{
		double Value;

		public static explicit operator Distance(double a) => new() { Value = a };
		public static explicit operator double(Distance a) => a.Value;
	}

I have read the [information page](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0075?f1url=%3FappId%3Droslyn%26k%3Dk(CS0075)), but still don't find it convincing on such simple and non-ambiguous cases.

**Comment by Youssef1313** [comment]
> int i = (System.Int32) -4;

It's ambiguous (syntactically) between subtraction and casting.

**Comment by IEvangelist** [comment]
Hi @ChrisBertrandIndus - thanks for posting this issue (and thanks @Youssef1313 as well). I'm curious, how exactly would you propose removing this error? The compiler raises this error when it is unable to disambiguate your intended code.

I do believe that this doc could be improved, here is a [Sharplab.io I threw together](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUBbAgARQ0C8NAFNQJQ0C0ALANw0CaAeiE0AxgHtkyAKZiEFEdRoArRiwCSMBAGZCHHv2GjYk6XIXllAa3WHBDwSJrAZCBDOQVFQgCJQAZwQIGDEZGgx1Zn8gkLC2TkIAVgA6AAYjAWdTKVl5Chjg0PDwpmjAovjmRNS0tiNnMzzLCnxiAE5mABIAInomAG8oAF84GltB61GaUpoBmWGe+u9yfB0aPKlIoLR5GkK4mQoBigBINYiJVGAAG3CANQgb1BleFfP1toA2GYAPAAcblAxFBaBJ/p4IAgpPsKodmBgrrdwhAOAwAHw0CiOQQwGQAd2YHAGNEez1mEBowze5A+NG+f0BwNBNHBkOhyEu1zu5VixRoqMYmOxOIFKTJLxpwyAA=) quick.

```csharp
using System;

int i = (int) -4;    // compiles
//int j = (Int32) -4;  // doesn't compile
int k = -4;          // compiles

//Distance d = (Distance)-25.0;    // doesn't compile
Distance e = (Distance)(-25.0);  // compiles

Console.WriteLine($"i = {i}, k = {k}, e = {e}");

public record struct Distance
{
	public double Value;

	public static explicit operator Distance(double a) => 
            new() { Value = a };
	public static explicit operator double(Distance a) => 
            a.Value;
}
```

The doc reads as:

> If you are casting using a keyword that identifies a predefined type, then you do not need parentheses. That's why this is fine, `int i = (int) -4;` but this is not `int j = (Int32) -4;`.

This part of the C# Specification, Section 7.6.6 and the doc says:

> Otherwise, you must put the parentheses because `(x) –y` will not be considered a cast expression.

How would you like it to read otherwise?

**Comment by RikkiGibson** [comment]
I wasn't able to find a Section 7.6.6 in the C# specification. Could you please link to the document which has such a section?

https://github.com/dotnet/csharpstandard/tree/draft-v6/standard

**Comment by ChrisBertrandIndus** [comment]
@Youssef1313 
Thank you for your answer.

I understand there is a difference between a negation and a substraction.

In `int i = (System.Int32) -4;`, I think the analyser should understand that [minus four] and [{nothing} less four] produce the same value : [-4]. Then it can cast it System.Int32.

Maybe that implies a different approach in the compilation process. But this ambiguity can easily lead to one interpretation only in such simple cases.
The fact is we, humans, interpret that expression as minus-four, not as a worthless substraction.

The same comes with a structure, such as `Distance d = (Distance)-25.0`, because of the explicit cast from an Double in the structure.

I fully understand a general-purpose analyser can run in separate steps, and that one of these steps is stuck on what it thinks is an ambiguity.
If it is such a problem, I suggest to think the analyser differently.

@IEvangelist 
I agree your clarification would improve the documentation, thank you for that.
My suggestion remains, even if that implies to rethink the analyser and not just adjust a few rules in it.

Thank you all anyway. I don't intend to insist much more on that suggestion.

**Comment by Youssef1313** [comment]
Let me try to elaborate more. Assume you have declared an enum called `E`, and have the following code

```csharp
    void M()
    {
        _ = (E) -4;
    }
```

Is this cast? or is `E` referring to a declared constant for example?

You could argue that it can be considered a cast if there is no constant or variable in scope with the name `E`, but the issue here is that it's not how the compiler works. I'll explain according to my understanding. But again, the compiler team may be able to provide a more in-depth explanation than me 😃 

The compiler works on different phases, the first of them lexing and parsing, which produces the syntax tree. This phase isn't aware of what variables or constants are in scope. It just represents "Syntax", and this phase has to produce a syntax tree.

The syntax you're referring to is ambiguous to this phase, is this subtracting a symbol called (`System.Int32`, ie a field/constant called `Int32` contained in a class called `System`)? or is this casting? Currently, the parser considers it as a subtraction.

The reason it works with the keyword "int" is that the parser can clearly tell that this is a reserved keyword, so it's 100% a cast.

Also keep in mind that the "same text" should always produce the "syntax tree". Let's (arbitrarily) assume the parser knows there is a variable in scope (which isn't how it works nor how it should work, this is an imaginary assumption). Then, adding a class in completely another file can generate a different syntax tree, which isn't desirable.
